File: brysbaert2014_Concreteness/preprocess_data.R
### Preprocess Data from: Brysbaert, M., Warriner, A.B., & Kuperman, V. (2014).
# Concreteness ratings for 40 thousand generally known English word lemmas.
# Behavior Research Methods, 46(3), 904–911.
# https://doi.org/10.3758/s13428-013-0403-5
#
# Raw data source: https://osf.io/qpmf4/
### Niklas Jung 
### 04/2026


library(dplyr)
library(readr)
library(tibble)
library(stringr)

# Resolve paths from this script's location so it runs from any working
# directory and always writes inside its own study folder.
.args <- commandArgs(trailingOnly = FALSE)
SCRIPT_DIR <- dirname(sub("^--file=", "", .args[grep("^--file=", .args)]))
if (length(SCRIPT_DIR) == 0 || !nzchar(SCRIPT_DIR)) SCRIPT_DIR <- getwd()


## read data 

# original data must be in same dir
stopifnot("original_data" %in% list.files(SCRIPT_DIR))

load(file.path(SCRIPT_DIR, "original_data", "concreteness_trial.rda"))


df_raw <- concreteness.participant

df <- df_raw %>%
  rename(
    participant_id     = Worker_code,        # anonymized ID — this is the real one
    age                = Answer.QAge,
    gender             = Answer.QGender,
    education          = Answer.QEducation,
    first_language     = Answer.QLang,
    handedness         = Answer.QHand,
    state_raised       = Answer.QRaised,     # US state before age 7
    stimulus           = Word,
    response           = Rating,
    session_duration_s = WorkTimeInSeconds
  ) %>%
  # Drop raw MTurk ID (privacy) and approval rate columns (infrastructure)
  select(-WorkerId,
         -LifetimeApprovalRate,
         -Last30DaysApprovalRate,
         -Last7DaysApprovalRate)


# Handle "n"/"N" = participant didn't know the word
df <- df %>%
  mutate(
    country_of_residence = "US", 
    word_known = if_else(response %in% c("n", "N"), FALSE, TRUE)
  )



# fix spelling differences in first_language (ALL have english as first language)
english_variants <- c(
  "English", "english", "ENGLISH", "Eglish", "ENglish", "Englsih",
  "English5", "English-", "eNGLISH", "Engish", "Englis", "Engllish",
  "nglish", "englidh", "Englisih", "Enlish", "englis", "Enlgish",
  "English,", "enlish", "Englishq", "American English", "ENGLLISH",
  "enlgish", "english5", "Englisg", "english ny", "english ph",
  "english californion", "englih", "Emglish", "Englosh", "Englishh",
  "eng", "Englilsh", "English (American)", "engllsh", "Engilsh",
  "\"English,", "Englis5", "american english", "Wnglish", "Elglish",
  "englosh"
)

df <- df %>%
  mutate(first_language = if_else(first_language %in% english_variants, 
                                  "English", 
                                  first_language))


df <- df %>%
  mutate(
    age = str_remove(age, "\\s*years?"),  # remove "years" / "year"
    age = as.numeric(age),
    age = if_else(age < 16 | age > 120, NA_real_, age)  # implausible values → NA
  )


### Add if teh word is normal trial or calibrator or control word

calibrator_words <- c("shirt", "Inf", "gas", "grasshopper", "marriage", 
                      "kick", "polite", "whistle", "theory", "sugar")

control_words <- c("adolescence", "aspect", "banality", "bottle", "cold",
                   "diving", "equanimity", "fishing", "fourth", "gnat",
                   "grisly", "harpsichord", "heroin", "impossible", "lathe",
                   "loquacity", "milk", "mist", "monument", "pen",
                   "pineapple", "saddle", "same", "science", "sigh",
                   "snap", "sock", "sympathy", "tumble")

df <- df %>%
  mutate(word_type = case_when(
    stimulus %in% calibrator_words ~ "calibrator",
    stimulus %in% control_words    ~ "control",
    TRUE                           ~ "normal"
  ))

# Add trial_order per participant
# -----------------------------------------------------------------------------
# The raw data has no explicit trial order column; we derive it from row
# position within each participant's session.


df <- df %>%
  arrange(participant_id) %>%
  group_by(participant_id) %>%
  mutate(
    session_id = ceiling(row_number() / 339)
  ) %>%
  ungroup()



#  Basic validation

#  Every row should have a stimulus
missing_stimulus <- sum(is.na(df$stimulus))
if (missing_stimulus > 0) {
  warning(missing_stimulus, " rows with missing stimulus. Inspect before proceeding.")
}

# 5c. Summary
message("\n--- Processed data summary ---")
message("Rows            : ", nrow(df))
message("Participants    : ", n_distinct(df$participant_id))
message("Unique stimuli  : ", n_distinct(df$stimulus))

## write df to processed_data

write_csv(df, file = file.path(SCRIPT_DIR, "processed_data", "exp1.csv"))


# write codebook as csv for this study
write_codebook <- function(base_dir) {
  codebook_path <- file.path(base_dir, "CODEBOOK.csv")
  
  if (file.exists(codebook_path)) return(invisible(NULL))
  
  rows <- tribble(
    ~column_name,        ~description,
    "participant_id",    "Anonymized participant ID (derived from Worker_code)",
    "age",               "Participant age in years (from Answer.QAge)",
    "gender",            "Self-reported gender: Female / Male / Unspecified (from Answer.QGender)",
    "education",         "Highest completed education level (from Answer.QEducation)",
    "first_language",    "Participant's self-reported language(s) (from Answer.QLang)",
    "handedness",        "Participant's dominant hand: Left / Right / Unspecified (from Answer.QHand)",
    "country_of_residence", "Country of residence (all US, state specified in next var",
    "state_raised",      "US state where participant lived most of the time before age 7 (from Answer.QRaised)",
    "session_duration_s","Time in seconds between accepting and submitting the MTurk assignment (from WorkTimeInSeconds)",
    "stimulus",          "Stimulus word presented to the participant (from Word)",
    "response",          "Concreteness rating 1-5, or N / n for unknown (see new variable word_known)",
    "word_known",        "FALSE if participant indicated they did not know the word (Rating was 'n' or 'N')",
    "word_type",         "Is trial part of control word or calibrator (that everyone gets)",
    "session_id",        "Participants could do the task multiple times. Each one is noted by session_id (max (90)."
  )
  
  write_csv(rows, codebook_path)
  message("CODEBOOK.csv written to: ", codebook_path)
}

# Usage:
write_codebook(SCRIPT_DIR)

File: guenther2023grammaticality/preprocess_data.py
import os
from pathlib import Path
import pandas as pd
import re



def ensure_processed_dir(base_dir: Path) -> Path:
    processed_dir = base_dir / "processed_data"
    processed_dir.mkdir(exist_ok=True)
    return processed_dir


def write_codebook(base_dir: Path) -> None:
    codebook_path = base_dir / "CODEBOOK.csv"
    if codebook_path.exists():
        return
    rows = [
        {"column_name": "participant_id", "description": "Anonymized participant ID"},
        {"column_name": "age", "description": "Participant age in years"},
        {"column_name": "trial_order", "description": "Trial order index (factorized from raw trial_order)"},
        {"column_name": "stimulus", "description": "Sentence string (from 'sentence')"},
        {"column_name": "response", "description": "Grammaticality judgment (e.g., 'c' / 'n')"},
        {"column_name": "rt", "description": "Response time in ms (from 'rt')"},
    ]
    pd.DataFrame(rows).to_csv(codebook_path, index=False)


def preprocess(base_dir: Path) -> None:
    original_path = base_dir / "original_data" / "LLMgrammaticality_humans.csv"
    processed_dir = ensure_processed_dir(base_dir)
    write_codebook(base_dir)

    df = pd.read_csv(original_path)

    # Rename / create canonical columns
    df = df.rename(columns={"sentence": "stimulus"})

    # Factorize trial_order to integers starting at 1
    # The raw export names this column trial_id; the corpus name for a
    # per-participant position is trial_order.
    if "trial_id" in df.columns:
        df["trial_order"] = pd.factorize(df["trial_id"])[0] + 1
        
    # make participant_id string
    df["participant_id"] = df["participant_id"].astype(str)
    
    # turn age to float
    df["age"] = df["age"].astype(float)
    
    # turn rt to float
    df["rt"] = df["rt"].astype(float)

    # Select and sort
    cols = ["participant_id", "age", "trial_order", "stimulus", "response", "rt"]
    
    df_out = df.loc[:, [c for c in cols if c in df.columns]].copy()
    df_out = df_out.sort_values(by=[c for c in ["participant_id", "trial_order"] if c in df_out.columns])

    # Write
    out_path = processed_dir / "exp1.csv"
    df_out.to_csv(out_path, index=False)


if __name__ == "__main__":
    preprocess(Path(__file__).parent.resolve())


